File: src/main/groovy/jokatu/components/config/InputDeserialisers.java
package jokatu.components.config;

import jokatu.game.input.Input;
import jokatu.game.input.InputDeserialiser;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;

import javax.annotation.PostConstruct;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.text.MessageFormat;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import static ophelia.exceptions.voidmaybe.VoidMaybe.wrap;
import static ophelia.exceptions.voidmaybe.VoidMaybeCollectors.merge;

@Configuration
@ComponentScan("jokatu.game")
public class InputDeserialisers {

	private static final Logger log = LoggerFactory.getLogger(InputDeserialisers.class);

	private Map<Class<? extends Input>, InputDeserialiser<?>> deserialiserMap = new HashMap<>();

	private final ApplicationContext applicationContext;

	@Autowired
	public InputDeserialisers(ApplicationContext applicationContext) {
		this.applicationContext = applicationContext;
	}

	@PostConstruct
	public void populateDeserialisers() throws Exception {
		applicationContext.getBeansOfType(InputDeserialiser.class).values().stream()
				.map(wrap(this::addDeserialiser))
				.collect(merge())
				.throwOnFailure();
		if (deserialiserMap.isEmpty()) {
			throw new Exception("No input deserialisers could be found.");
		}

		log.debug("{} initialised", InputDeserialisers.class.getSimpleName());
	}

	@Nullable
	public InputDeserialiser<?> getDeserialiser(@NotNull Class<? extends Input> inputClass) {
		return deserialiserMap.get(inputClass);
	}

	private void addDeserialiser(@NotNull InputDeserialiser inputDeserialiser) throws Exception {
		Class<? extends Input> inputType = getInputType(inputDeserialiser);
		if (isAbstract(inputType)) {
			throw new Exception(MessageFormat.format(
					"The deserialiser {0} returns the abstract input class {1}",
					inputDeserialiser.getClass().getSimpleName(),
					inputType.getSimpleName()
			));
		}
		if (deserialiserMap.containsKey(inputType)) {
			throw new Exception(MessageFormat.format("The input class {0} has multiple deserialisers", inputType.getSimpleName()));
		}
		deserialiserMap.put(inputType, inputDeserialiser);
	}

	private boolean isAbstract(@NotNull Class<?> inputType) {
		return (inputType.getModifiers() & Modifier.ABSTRACT) == Modifier.ABSTRACT;
	}

	@NotNull
	private Class<? extends Input> getInputType(@NotNull InputDeserialiser deserialiser) throws Exception {
		Set<Class<?>> deserialiseReturnClasses = Arrays.stream(deserialiser.getClass().getDeclaredMethods())
				.filter(method -> !Input.class.equals(method.getReturnType()))
				.filter(method -> Input.class.isAssignableFrom(method.getReturnType()))
				.map(Method::getReturnType)
				.collect(Collectors.toSet());
		if (deserialiseReturnClasses.size() == 1) {
			//noinspection unchecked
			return (Class<? extends Input>) deserialiseReturnClasses.iterator().next();

		} else if (deserialiseReturnClasses.size() > 1) {
			throw new Exception(MessageFormat.format(
					"The deserialiser {0} has multiple declared methods that have a return type of a strict subclass of {1}.",
					deserialiser.getClass().getSimpleName(),
					Input.class.getSimpleName()
			));
		} else {
			throw new Exception(MessageFormat.format(
					"The deserialiser {0} has no declared methods that have a return type of a strict subclass of {1}.",
					deserialiser.getClass().getSimpleName(),
					Input.class.getSimpleName()
			));
		}
	}
}

File: src/main/kotlin/jokatu/components/config/InputDeserialisers.kt
package jokatu.components.config

import jokatu.game.input.Input
import jokatu.game.input.InputDeserialiser
import ophelia.exceptions.voidmaybe.VoidMaybe.wrap
import ophelia.exceptions.voidmaybe.VoidMaybeCollectors.merge
import ophelia.function.ExceptionalConsumer
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.getBeansOfType
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.Configuration
import reflect.asSubclass
import java.util.*
import java.util.stream.Collectors
import javax.annotation.PostConstruct
import kotlin.reflect.KClass
import kotlin.reflect.full.declaredMemberFunctions
import kotlin.reflect.full.isSubclassOf

@Configuration
@ComponentScan("jokatu.game")
class InputDeserialisers
@Autowired constructor(
		private val applicationContext: ApplicationContext
) {
	private val deserialiserMap = HashMap<KClass<out Input>, InputDeserialiser<*>>()

	@PostConstruct
	@Throws(Exception::class)
	fun populateDeserialisers() {
		applicationContext.getBeansOfType(InputDeserialiser::class).stream()
				.map(wrap(ExceptionalConsumer<InputDeserialiser<*>, Exception> { this.addDeserialiser(it) }))
				.collect(merge())
				.throwOnFailure()
		if (deserialiserMap.isEmpty()) {
			throw Exception("No input deserialisers could be found.")
		}

		log.debug("{} initialised", InputDeserialisers::class.java.simpleName)
	}

	fun getDeserialiser(inputClass: KClass<out Input>): InputDeserialiser<*>? {
		return deserialiserMap[inputClass]
	}

	@Throws(Exception::class)
	private fun addDeserialiser(inputDeserialiser: InputDeserialiser<*>) {
		val inputType = getInputType(inputDeserialiser)
		if (inputType.isAbstract) {
			throw Exception(
					"The deserialiser ${inputDeserialiser.javaClass.simpleName} returns the abstract input class "
							+ "${inputType.simpleName}."
			)
		}
		if (deserialiserMap.containsKey(inputType)) {
			throw Exception("The input class ${inputType.simpleName} has multiple deserialisers.")
		}
		deserialiserMap[inputType] = inputDeserialiser
	}

	@Throws(Exception::class)
	private fun getInputType(deserialiser: InputDeserialiser<*>): KClass<out Input> {
		val deserialiseReturnClasses = deserialiser::class.declaredMemberFunctions.stream()
				.map { method -> method.returnType.classifier }
				.filter { clazz -> clazz is KClass<*> }
				.map { clazz -> clazz as KClass<*> }
				.filter { clazz -> clazz.isSubclassOf(Input::class) && clazz != Input::class }
				.map { clazz -> clazz.asSubclass(Input::class) }
				.collect(Collectors.toSet())
		when {
			deserialiseReturnClasses.size == 1 -> return deserialiseReturnClasses.first()
			deserialiseReturnClasses.size > 1 -> throw Exception(
					"The deserialiser ${deserialiser::class.simpleName} has multiple declared methods that have a "
							+ "return type of a strict subclass of ${Input::class.simpleName}."
			)
			else -> throw Exception(
					"The deserialiser ${deserialiser::class.simpleName} has no declared methods that have a return "
							+ "type of a strict subclass of ${Input::class.simpleName}."
			)
		}
	}

	companion object {
		private val log = LoggerFactory.getLogger(InputDeserialisers::class.java)
	}
}
